File: Solutions/812-rotate-string/rotate-string.swift
class Solution {
    func rotateString(_ s: String, _ goal: String) -> Bool {
        // Step 1: If lengths are different, return false
        if s.count != goal.count {
            return false
        }
        
        // Step 2: Concatenate s with itself
        let doubleS = s + s
        
        // Step 3: Check if goal is a substring of doubleS
        return doubleS.contains(goal)
    }
}

File: Solutions/812-rotate-string/rotate-string.cs
public class Solution {
    public bool RotateString(string s, string goal) {
        
        // -------------------------------
        // STEP 1: LENGTH CHECK
        // -------------------------------
        // If lengths are different, it's impossible
        // to form goal by rotating s
        if (s.Length != goal.Length) {
            return false;
        }

        // -------------------------------
        // STEP 2: CONCATENATE STRING
        // -------------------------------
        // Create a new string that contains all possible rotations
        // Example: "abcde" -> "abcdeabcde"
        string doubled = s + s;

        // -------------------------------
        // STEP 3: CHECK SUBSTRING
        // -------------------------------
        // If goal is a rotation of s,
        // it must appear inside doubled string
        if (doubled.Contains(goal)) {
            return true;
        }

        // Otherwise, not a valid rotation
        return false;
    }
}